Lib/urllib/parse.py

```diff
@@ -683,9 +692,21 @@ def parse_qsl(qs, keep_blank_values=False, strict_parsing=False,
         encoding and errors: specify how to decode percent-encoded sequences
             into Unicode characters, as accepted by the bytes.decode() method.
 
+        max_num_fields: Integer. If set, then throws an ValueError
+            if there are more than n fields read by parse_qsl().
+
         Returns a list, as G-d intended.
     """
     qs, _coerce_result = _coerce_args(qs)
+
+    # If max_num_fields is defined then check that the number of fields
+    # is less than max_num_fields. This prevents a memory exhaustion DOS
+    # attack via post bodies with many fields.
+    if max_num_fields is not None:
+        for num_fields, _ in enumerate(_QS_DELIMITER_RE.finditer(qs), 2):
```

Comment:
Doing this with `finditer()` instead of `count()` so the worst case time is bounded by `max_num_fields` instead of `len(qs)`.

Reply:
I feel `qs.count('&') + qs.count(';')` is much simpler and faster.
Is this complexity really needed?

When qs is `'a'*1000 + ';&'`, finditer()'s time is bounded by `len(qs)`

Reply:
@methane Let me know what you think of that explanation and I'll make the switch if you want, whichever way is okay with me.

Reply:
Receiving 9MB query string is slow already.  I don't think "500x slower than finditer" is not a problem in such case.

There are other methods:

```py
pairs = [s2 for s1 in qs.split('&', max_num_fields) for s2 in s1.split(';', max_num_fields)]
if len(pairs) > max_num_fields:  # len(pairs) may be `(max_num_fields+1)*2`.  But since it's not cgi.FieldStorage, it's not a problem.
    raise ValueError('Max number of fields exceeded')
```

Or you can use the regex to make pairs too:

```py
pairs = _QS_DELIMITER_RE.split(qs, max_num_fields)
if len(pairs) > max_num_fields:
    raise ValueError('Max number of fields exceeded')
```

I think these are simpler and faster than `enumerate` hack.

Reply:
Alright, thanks for taking a look @methane . I changed that to use `count()` which is simplest in commit 1fa59e4efd. If that looks good can you add the backport labels so that the [miss-islington bot](https://devguide.python.org/committing/#backporting-changes-to-an-older-version) automatically backports this? I expect the automatic backports to fail on 2.7 but I'll fix that manually. I intend to backport it to all maintained versions, which I believe are 3.5-3.7 and 2.7.